spring-xd-dirt/src/main/java/org/springframework/xd/dirt/container/decryptor/PropertiesDecryptor.java

```diff
@@ -0,0 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.xd.dirt.container.decryptor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.context.event.ApplicationPreparedEvent;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.core.Ordered;
+import org.springframework.core.env.CompositePropertySource;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.core.env.MapPropertySource;
+import org.springframework.core.env.MutablePropertySources;
+import org.springframework.core.env.PropertySource;
+import org.springframework.core.env.PropertySources;
+import org.springframework.security.crypto.encrypt.TextEncryptor;
+import org.springframework.xd.dirt.container.initializer.OrderedContextInitializer;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Decrypt properties from the environment and insert them with high priority so they
+ * override the encrypted values. This class is injected as an ApplicationListener in
+ * each application context in the XD hierarchy to be invoked before the context is
+ * refreshed. This class is loosely based on spring-cloud-commons
+ * org.springframework.cloud.bootstrap.encrypt.EnvironmentDecryptApplicationInitializer,
+ * adopting the same conventions used to identify and process encrypted property values
+ * in Spring Cloud applications.
+ *
+ * @author David Turanski
+ * @since 1.3.1
+ */
+public class PropertiesDecryptor implements OrderedContextInitializer {
+	private static Logger logger = LoggerFactory.getLogger(PropertiesDecryptor.class);
+
+	public static final String DECRYPTED_PROPERTY_SOURCE_NAME = "decrypted";
+
+	public static final String DECRYPTOR_BEAN_NAME = "propertiesDecryptor";
+
+	private int order = Ordered.HIGHEST_PRECEDENCE + 15;
+
+	private TextEncryptor decryptor;
+
+	private boolean failOnError = true;
+
+	/**
+	 * Strategy to determine how to handle exceptions during decryption.
+	 *
+	 * @param failOnError the flag value (default true)
+	 */
+	public void setFailOnError(boolean failOnError) {
+		this.failOnError = failOnError;
+	}
+
+	public PropertiesDecryptor(TextEncryptor decryptor) {
+		this.decryptor = decryptor;
+	}
+
+	@Override
+	public int getOrder() {
+		return this.order;
+	}
+
+	public void setOrder(int order) {
+		this.order = order;
+	}
+
+	@Override
+	public void onApplicationEvent(ApplicationPreparedEvent event) {
+		if (this.decryptor == null) {
+			return;
+		}
+
+		ConfigurableApplicationContext applicationContext = event
+				.getApplicationContext();
```

Comment:
Needs `ApplicationContextAware` and check that the event is for this bean's context.